modules/pico_fragments.c

```diff
@@ -430,7 +430,14 @@ void pico_ipv6_process_frag(struct pico_ipv6_exthdr *frag, struct pico_frame *f,
     if (first)
     {
       if ((pico_ipv6_frag_match(f, first) == 0 && (IP6_FRAG_ID(frag) == ipv6_cur_frag_id))) {
-        pico_tree_insert(&ipv6_fragments, pico_frame_copy(f));
+        if(pico_tree_insert(&ipv6_fragments, pico_frame_copy(f)) ){
+			frag_dbg("Could not insert picoframe in tree\n");
+			if(pico_err != PICO_ERR_ENOMEM){
+				frag_dbg("Key is already in tree\n");
+				pico_err = PICO_ERR_EINVAL;
+			}
+			return;
```

Comment:
Shouldn't you delete the result of the `pico_frame_copy` here? (same goes for other occurrences in this file)
And actually, the frame copy could go wrong as well (also due to memory issues), so you might want to check for that as well, before trying to insert.